core/events/__init__.py

```diff
@@ -259,8 +259,7 @@ def _apply_events(events: List[Event], rules: List[EventRule],
 
     extra_events: List[Event] = []    # Generated by applied rules.
     for event in events:
-        if not event.valid(submission):
-            raise InvalidEvent(event)
+        event.validate(submission)      # Will throw InvalidEvent.
```

Comment:
This was creating some unnecessary indirection, so I changed it to let the InvalidEvent percolate up from the event's ``validate`` method. Simpler, easier to debug.